kaitai/kaitaistream.h

```diff
@@ -261,6 +270,25 @@ class kstream {
     static const int ZLIB_BUF_SIZE = 128 * 1024;
 };
 
+template <class From>
+//inline typename std::make_signed<From>::type to_signed(From from) {
+inline auto to_signed(From from) {
+    using signed_F = typename std::make_signed<From>::type;
+
+    if (from <= std::numeric_limits<signed_F>::max()) {
+        return static_cast<signed_F>(from);
+    }
+
+    throw std::runtime_error("toSigned: Invalid casting");
```

Comment:
@GreyCat @maks500 
Unfortunately, this line is currently causing the whole C++11 CI build on Linux to fail (https://travis-ci.org/kaitai-io/ci_targets/jobs/660553811#L584):

```cpp
/home/travis/build/kaitai-io/ci_targets/tests/spec/cpp_stl_11/../../../runtime/cpp_stl/kaitai/kaitaistream.h:282:11: error: ‘runtime_error’ is not a member of ‘std’
     throw std::runtime_error("toSigned: Invalid casting");
```

Another issue is that it's a bad idea to throw generic runtime exceptions, it's generally recomended to always create specific exceptions which can be handled much easier from the application code.

Reply:
> ‘runtime_error’ is not a member of ‘std’

Actually, it _is_ a member of `std`, so perhaps there's a missing header file or similar issue?

> it's a bad idea to throw generic runtime exceptions

I would suggest that that depends on the situation.  If the intention is that the caller be able to recover from the exception, then the I concur that the API should include defined exception types or objects which convey to the caller all the information required to handle the situation when one is thrown; however, if the expectation is that there is no recovery available, then there is likely no reason to create anything fancier than one of the "standard" exception objects -- all that is required is that a human be able to diagnose the problem after the fact.  (As such, I'd say that this code looks about right, although it might be nice if the message included a little more detail, such as the location in the KSY where the offending cast can be found.)

Reply:
> however, if the expectation is that there is no recovery available, then there is likely no reason to create anything fancier than one of the "standard" exception objects -- all that is required is that a human be able to diagnose the problem after the fact.

I agree, but if this is the case, then I'd say it's not a valid use case for `runtime_error` from the semantic POV. There are generally two groups of exceptions: **runtime** and **logic** ones. Runtime errors can be found only on runtime occurs, whereas logic errors _represent error in the program logic and are _presumably detectable before the program executes_ ([source](http://www.cplusplus.com/reference/stdexcept/logic_error/)). That means that logic errors are not intended to be caught by the application code, so there is no point creating some specific subclasses.

So if this error has logic nature (as you denote), then it should be [`std::logic_error`](http://www.cplusplus.com/reference/stdexcept/logic_error/). But  **runtime errors** (for example end-of-stream errors are typical runtime ones - you can't detect them by just inspecting the code because they can be triggered just by some specific user input) should IMHO **always** have specific subclasses, because they're intended to be caught by the application code.

Reply:
I suspect that we are going to be challenged by terminology and context here.  :-)

I believe that there are in fact _two_ run times in play:  there is the operation of the Kaitai Struct Compiler, and there is the operation of the code that it produced.  Concordantly, there are two sets of "user input", respectively:  the KSY file, and the to-be-parsed data.

If the exception is raised during the execution of KSC in response to some aspect of the KSY, then I think `std::runtime_error` is reasonable.  If the exception is raised during the execution of the resulting application's call to the KSC-generated code, then things are murkier.  (I apologize for not having a good enough understanding of this code change to be able to tell where we are....)

In the latter case, I don't particularly like the use of `std::runtime_error`, but I think it's acceptable; however, I don't particularly like the use of `std::logic_error` either.  What we're trying to report is, presumably, an error which occurred during the construction of the application due to a feature of the KSY, but which wasn't detected until a execution of the application.  In any case, there is nothing that the application can do about it at run-time, and so I don't think a subclass (etc.) is warranted.

(And, I disagree with your assertion that just because something is a "runtime error" means that it is something that the application can or is intended to be able to recover from -- some errors are "fatal" -- but they can and should be reported via exceptions to allow the callers to restore any external invariants prior to halting execution.  If they are detected at run-time and don't represent an error in the internal logic of the program, there is no reason not to report them as `std::runtime_error`'s.)